docs/docs/higher-order-components.md

```diff
@@ -0,0 +1,491 @@
+
+```js
+// Instead of doing this...
+const EnhancedComponent = connect(commentSelector)(withRouter(WrappedComponent))
+
+// ... you can use a function composition utility
+// compose(f, g, h) is the same as (...args) => f(g(h(...args)))
+const enhance = compose(
+  // These are both single-argument HOCs
+  connect(commentSelector),
+  withRouter
+)
+const EnhancedComponent = enhance(WrappedComponent)
+```
+
+(This same property also allows `connect` and other enhancer-style HOCs to be used as decorators, an experimental JavaScript proposal.)
+
+The `compose` utility function is provided by many third-party libraries including lodash (as [`lodash.flowRight`](https://lodash.com/docs/#flowRight)), [Redux](http://redux.js.org/docs/api/compose.html), and [Rambda](http://ramdajs.com/docs/#compose).
+
+### Optional: Wrap the Display Name for Easy Debugging
+
+The container components created by HOCs show up in the React DevTools like any other component. To ease debugging, choose a display name that communicates that it's the result of an HOC.
+
+The most common technique is to wrap the display name of the wrapped component. So if your higher-order component is named `withSubscription`, and the wrapped component's display name is `CommentList`, use the display name `WithSubscription(CommentList)`:
+
+```js
+function withSubscription(WrappedComponent) {
+  class WithSubscription extends React.Component {/* ... */}
+  WithSubscription.displayName = `WithSubscription(${getDisplayName(WrappedComponent)})`;
+  return WithSubscription;
+}
+
+function getDisplayName(WrappedComponent) {
+  return WrappedComponent.displayName || WrappedComponent.name || 'Component';
+}
+```
+
+
+## Caveats
+
+We consider HOCs an advanced technique because they come with a few caveats that aren't immediately obvious if you're new to React.
+
+### Don't Use HOCs Inside the render Method
+
+React's diffing algorithm (called reconciliation) uses component identity to determine whether it should update the existing subtree or throw it away and mount a new one. If the component returned from `render` is identical (`===`) to the component from the previous render, React recursively updates the subtree by diffing it with the new one. If they're not equal, the previous subtree is unmounted completely.
+
+Normally, you shouldn't need to think about this. But it matters for HOCs because it means you can't apply an HOC to a component within the render method of a component:
+
+```js
+render() {
+  // A new version of EnhancedComponent is created on every render
+  // EnhancedComponent1 !== EnhancedComponent2
+  const EnhancedComponent = enhance(MyComponent);
+  // That causes the entire subtree to unmount/remount each time!
+  return <EnhancedComponent />;
+}
+```
+
+The problem here isn't just about performance—remounting a component causes the state of that component and all of its children to be lost.
```

Comment:
I think React docs use the style with spaces around em dash. I allow myself to do my own thing in blog posts but let's be consistent in docs and put spaces there I guess